samcli/lib/sync/flows/function_sync_flow.py

```diff
@@ -98,5 +101,27 @@ def gather_dependencies(self) -> List[SyncFlow]:
 
         return sync_flows
 
+    def _verify_function_status(self) -> bool:
+        """
+        return value:
+        True if last function update was successful
+        False if last function update was still in progress or failed
+        """
+        response = self._lambda_client.get_function(FunctionName=self.get_physical_id(self._function_identifier))
+        if response.get("Configuration", {}).get("LastUpdateStatus", "") == FunctionUpdateStatus.IN_PROGRESS.value:
+            time.sleep(FUNCTION_SLEEP)
+            return self._verify_function_status()
```

Comment:
nit: This is designed as recursive calls, should we change this to a while loop? Python has 1000 recursive call limit by default, I don't think we are going to exceed that, but just checking whether it would be safer or not.
```
>>> import sys
>>> sys.getrecursionlimit()
1000
```

Reply:
The reason why while loop is not preferred is that it needs to return a tuple of booleans in order to tell apart Success, InProgress and Failed. Something like
```
while in_progress and not failed:
  in_progress, not failed = verify_function_status()
```
 would work, but I feel that a recursion is cleaner.
 I can add in a max retry number into this recursion that if after 100 retries doesn't if it is still in progress then we just continue. Then the expected exception would be race condition, which is better than max recursion reached.